crates/sui-json-rpc/src/coin_api.rs

```diff
@@ -288,41 +209,706 @@ impl CoinReadApiServer for CoinReadApi {
-
-            let metadata_object = self
-                .find_package_object(
-                    &coin_struct.address.into(),
-                    CoinMetadata::type_(coin_struct),
-                )
-                .await
-                .ok();
-
-            Ok(metadata_object.and_then(|v: Object| v.try_into().ok()))
+            let coin_struct = parse_sui_struct_tag(&coin_type).map_err(|_| {
+                Error::SuiRpcInputError(SuiRpcInputError::CannotParseSuiStructTag(coin_type))
+            })?;
+            Ok(self.internal.get_coin_metadata(coin_struct).await?)
         })
     }
 
     #[instrument(skip(self))]
     async fn get_total_supply(&self, coin_type: String) -> RpcResult<Supply> {
         with_tracing!(async move {
-            let coin_struct = parse_sui_struct_tag(&coin_type)?;
+            let coin_struct = parse_sui_struct_tag(&coin_type).map_err(|_| {
+                Error::SuiRpcInputError(SuiRpcInputError::CannotParseSuiStructTag(coin_type))
+            })?;
 
             Ok(if GAS::is_gas(&coin_struct) {
                 Supply { value: 0 }
             } else {
-                let treasury_cap_object = self
-                    .find_package_object(
-                        &coin_struct.address.into(),
-                        TreasuryCap::type_(coin_struct),
-                    )
-                    .await?;
-
-                let treasury_cap = TreasuryCap::from_bcs_bytes(
-                    treasury_cap_object.data.try_as_move().unwrap().contents(),
-                )
-                .map_err(Error::from)?;
+                let treasury_cap = self.internal.get_treasury_cap(coin_struct).await?;
                 treasury_cap.total_supply
             })
         })
     }
 }
+
+#[cfg_attr(test, automock)]
+#[async_trait]
+pub trait StateServiceWrapper {
+    fn get_object_read(&self, object_id: &ObjectID) -> SuiResult<ObjectRead>;
+    async fn get_object(&self, object_id: &ObjectID) -> SuiResult<Option<Object>>;
+    fn find_publish_txn_digest(&self, package_id: ObjectID) -> SuiResult<TransactionDigest>;
+    async fn find_package_object(
+        &self,
+        package_id: &ObjectID,
+        object_struct_tag: StructTag,
+    ) -> SuiApiResult<Object>;
+    fn get_owned_coins(
+        &self,
+        owner: SuiAddress,
+        // If `Some`, the query will start from the next item after the specified cursor
+        cursor: (String, ObjectID),
+        limit: usize,
+        one_coin_type_only: bool,
+    ) -> SuiResult<Vec<SuiCoin>>;
+    async fn get_executed_transaction_and_effects(
+        &self,
+        digest: TransactionDigest,
+    ) -> SuiResult<(VerifiedTransaction, TransactionEffects)>;
+    async fn get_balance(&self, owner: SuiAddress, coin_type: TypeTag) -> SuiResult<TotalBalance>;
+    async fn get_all_balance(
+        &self,
+        owner: SuiAddress,
+    ) -> SuiResult<Arc<HashMap<TypeTag, TotalBalance>>>;
+}
+
+pub struct AuthorityStateWrapper {
```

Comment:
Do you need this wrapper type, or could you implement your trait directly on `Arc<AuthorityState>` (or on `AuthorityState` itself?)

Reply:
Oh I misunderstood, you meant something like `impl Trait for AuthorityState` right? Yeah that makes a lot of sense too